tuned/exports/unix_socket_exporter.py

```diff
@@ -0,0 +1,228 @@
+			self._socket_object.close()
+
+	def _send_data(self, s, address, type, data):
+		"""
+		Sends data via socket s to address.
+		First byte of message specifies type of message (currently "r" for response, "e" for error message and "s" for signal).
+		Four more bytes are integer in network/big-endian with length of rest of message.
+		Rest of message is json data.
+		"""
+		log.debug("Sending socket data (type: %s, data: %s) to %s" % (type, data, address))
+		try:
+			j = json.dumps(data).encode("utf-8")
+			s.sendto(struct.pack("!sL", type.encode("utf-8"), len(j)), address)
+			s.sendto(j, address)
+		except Exception as e:
+			log.warning("Failed send data '%s' to '%s' type '%s': %s" % (data, address, type, e))
+
+	def period_check(self):
+		"""
+		Periodically checks socket object for new calls. This allows to function without special thread.
+		First four bytes is expected to be integer in network/big-endian with length of rest of message.
+		Rest of message expects json array, where first item is name of called method, rest is arguments of given method.
+		For response see method _send_data.
+		
+		Example implementation:
+		
+		def sss(d):
+			ssock_file = "/tmp/tuned_socket"
+			csock_file = "/tmp/tuned_socket_client"
+			if os.path.exists(csock_file):
+				os.remove(csock_file)
+			s=socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+			s.bind(csock_file)
+			j = json.dumps(d).encode("utf-8")
+			try:
+				s.sendto(struct.pack("!L", len(j)), ssock_file)
+				s.sendto(j, ssock_file)
+				d = s.recvfrom(5)
+				t, l = struct.unpack("!sL", d[0])
+				print("TYPE: %s" % t)
+				print("LEN: %s" % l)
+				data = json.loads(s.recvfrom(l)[0])
+				print("DATA: %s" % data)
+			except Exception as e:
+				print("E: %s" % e)
+			s.close()
+			
+		sss(("profiles",))
+		sss(("switch_profile", "balanced"))
+		"""
+		r, _, _ = select.select([self._socket_object], (), (), 0)
+		if r:
+			try:
+				l, address = self._socket_object.recvfrom(BYTES_LEN_OF_MSG)
+				l = struct.unpack("!L", l)[0]
+				log.debug("%s %s" % (address, l))
+			except Exception as e:
+				log.error("Failed to load length of message: %s" % e)
+				return
+			try:
+				data, address = self._socket_object.recvfrom(l)
+				log.debug("%s %s" % (address, data))
+			except Exception as e:
+				log.error("Failed to load data of message: %s" % e)
+				return
+			if not address:
+				log.warning("No return address for event, skipping")
+				return
+			if data:
+				try:
+					log.debug(data.decode("utf-8"))
+					data = json.loads(data.decode("utf-8"))
+				except Exception as e:
+					log.error("Failed to load json data '%s': %s" % (data, e))
+					self._send_data(self._socket_object, address, "err", "Failed to load json data '%s': %s" % (data, e))
+					return
+				if type(data) not in (tuple, list) or len(data) < 1:
+					log.error("Wrong format of call")
+					self._send_data(self._socket_object, address, "err", "Wrong format of call")
+				elif data[0] not in self._unix_socket_methods:
```

Comment:
Multiple methods are not really needed at long as there is a single method that does this :) But you are right that we will typically do a hotunplug / hotplug together (moving a CPU from one section to another).

Reply:
Coming back to this subject as I believe API is probably the most important to get right.  Code can be fixed afterwards, API not so easily...

I suggest the following (yaml is easier to read, please convert to json on your own):
```
version: 0.0.1
methods:
- args:
  - arg1
  - arg2
  - arg3
  command: method1
- args:
  - arg1
  - arg2
  - arg3
  command: method2
```

or even simpler

```
version: 0.0.1
methods:
- command:
  - name
  - arg1
  - arg2
- command:
  - name
  - arg1
  - arg2
```

We can even drop `version` and add it later if/when needed, but as the `json` is structured now, it will be impossible to change/enhance in the future.  I'm not even suggesting to implement more than one method.  That can even be done later if it is too difficult to implement (would be nice though), but I belive the json structure has to change to be more flexible.  Thoughts? @MarSik , @yarda  I'm not good with naming things, feel free to suggest better names and structure.

Reply:
Please please do not reinvent the wheel. I mentioned it before: jsonrpc supports batches - https://www.jsonrpc.org/specification#batch

Reply:
@MarSik Do you want to implement similar structure to jsonrpc or is it ok as it's now? I f you have any suggestions, we can discuss it on other "faster" channels like IRC

Reply:
I suggest the `#tuned` IRC channel if/when Martin has time.  IMO the current implementation is not OK as it does not allow future enhancements easily enough.